### PR TITLE
implement the HMC kernel

### PR DESCRIPTION
**Pull request by rlouf**
## Scope

Implement a HMC kernel that moves one chain for one step, with the following elements:

- [x] Velocity verlet integrator + tests
- [x] Proposal with a fixed number of steps
- [x] Gaussian euclidean metric
- [x] HMC base kernels
- [x] High-level API
- [x] Setup simple integration test

The goal is to get to a concrete implementation quickly to see the design in action.

### CODE REVIEW

File: blackjax/hmc.py
Comment (twolodzko) [inline blackjax/hmc.py:13]: Technically `np.ndarray` is the type.

File: blackjax/hmc.py
Comment (twolodzko) [inline blackjax/hmc.py:14]: Those types are repeated, probably nicer to have them in single place for consistency.

File: blackjax/hmc.py
Comment (rlouf) [reply blackjax/hmc.py:13]: Technically you're right 🙂 Thanks!

File: blackjax/hmc.py
Comment (rlouf) [reply blackjax/hmc.py:14]: I agree. I usually duplicate things until where that single place should be becomes obvious.

**Comment by rlouf** [comment]
The HMC algorithm with diagonal mass matrix works. I still need to fix the dense matrix case.

**Review by rlouf — COMMENTED** [review]
I fixed the dense mass matrix case, and replaced `tree_flatten` with `ravel_pytree` (which also flattens arrays at the leaves). Ready for review.

File: blackjax/inference/base.py
Comment (MarcoGorelli) [inline blackjax/inference/base.py:63]: I don't know how far you want to go with type hints, but perhaps they could be added here as well? e.g.
```
    proposal_generator: Callable[[Array, HMCProposalState], Tuple[HMCProposalState, HMCProposalInfo]],
```
(if I've understood correctly)

(I can work on adding these if you're open to them)

Comment (rlouf) [reply blackjax/inference/base.py:63]: I think it could eventually be useful to users to have their IDE raise a warning if the function does not have the correct signature. Let's wait  until NUTS to see if the signature stays the same :)

Comment (MarcoGorelli) [reply blackjax/inference/base.py:63]: OK, sure!

**Comment by rlouf** [comment]
_Note for future self_

Implementing NUTS I realize there is a much better way to cut HMC. Indeed, HMC first generates a trajectory then draws either the last or first state at random. This can be generalized, as we may want to draw an intermediate state of the trajectory instead (see Conceptual Introduction to HMC). Or draw several states from this trajectory (see Recycled Hamiltonian Monte Carlo). The proposal is thus a combination of constructing the trajectory and sampling states from that trajectory. We should thus remove the metropolis acceptance step from `base_hmc` and have `hmc_proposal` be a combination of `forward_trajectory` and `choosing_start_or_end` sampling scheme. NUTS fits more easily in this formulation, and it allows to build all sorts of variants.

**Comment by twiecki** [comment]
Is this ready to merge?

**Comment by rlouf** [comment]
Ready to merge on my end, I was just waiting for potential reviews.

File: blackjax/inference/proposals.py
Comment (fehiepsi) [inline blackjax/inference/proposals.py:65]: I think this is useful but memory cost will be O(num_steps), while in NUTS, it is O(log(num_steps)) and in usual HMC, it is O(1).

**Review by fehiepsi — COMMENTED** [review]
Thanks @rlouf! Looks great overall. I just have some comments below.

File: blackjax/inference/base.py
Comment (neerajprad) [inline blackjax/inference/base.py:156]: We can avoid this PE evaluation by caching `potential_energy` or `log_prob` in `HMCProposalState`.

File: blackjax/inference/base.py
Comment (neerajprad) [inline blackjax/inference/base.py:177]: nit: If we store `potential_energy` in `HMCProposalState`, we can probably have a single `HMCState` for both representing `(q, p, potential_energy, potential_energy_grad)` and not have to store `proposal` separately in `HMCInfo`. Please disregard my comment if there are other reasons for this separation.

**Comment by neerajprad** [comment]
The interface looks nice to me! Feel free to disregard/stash some of my comments as future work, and merge this if needed.

File: blackjax/inference/integrators.py
Comment (rlouf) [reply blackjax/inference/integrators.py:53]: No it's not. It's a remnant of MCX where several palindromic integrators (cf https://arxiv.org/abs/1711.05337) are implemented, including one where `a1` is not null. I'd leave it as "documentation".

File: blackjax/inference/integrators.py
Comment (rlouf) [reply blackjax/inference/integrators.py:76]: You're right. In the NUTS PR I made the integrator return both the value of the potential. I'll introduce the change here too. Also `s/log_prob/potential_energy`.

File: blackjax/inference/integrators.py
Comment (rlouf) [reply blackjax/inference/integrators.py:58]: I'd expect it to be negligible compared to the gradient of the potential, but that's a good point. Maybe we should do some perf test on a separate branch once this one is merged?

File: blackjax/inference/integrators.py
Comment (rlouf) [reply blackjax/inference/integrators.py:64]: Yep there's a confusion in the names, it is the name. I will replace `log_prob` by `potential_enegy` as I said earlier.

File: blackjax/inference/proposals.py
Comment (rlouf) [reply blackjax/inference/proposals.py:65]: I do not have a case for this variable, apart from maybe diagnosis. With hindsight I believe we should discard all intermediate states by default; if someone ever needs the entire trajectories this can be implemented as a separate proposal.

Related to this I would like to implement a "proposal" that recycles intermediate states (https://arxiv.org/abs/1511.06925), but it can implemented so as to only keep the useful steps as well, discarding the ones that won't be in the trace.

File: blackjax/inference/metrics.py
Comment (rlouf) [reply blackjax/inference/metrics.py:46]: That's the constraint indeed, and I agree users should not have to read JAX's code to know that it does a `sort` on the dictionary keys when flattening :) I will add a short paragraph about that.

File: blackjax/hmc.py
Comment (rlouf) [reply blackjax/hmc.py:87]: Yes, do you see any reason not to? The current plan is to publish a first release with the HMC and NUTS kernel and then move on to window adaptation.

File: blackjax/inference/base.py
Comment (rlouf) [reply blackjax/inference/base.py:156]: Yes, you and @fehiepsi are right, I will edit the integrator.

File: blackjax/inference/base.py
Comment (rlouf) [reply blackjax/inference/base.py:177]: The `proposal` in `HMCInfo` is the last state of the trajectory. `new_state` is what I now think is what should be called the proposal. I return the last state of the trajectory in `HMCInfo` because it can be useful to diagnose when there are divergences or the acceptance rate is very low.

File: blackjax/inference/integrators.py
Comment (rlouf) [reply blackjax/inference/integrators.py]: Pushed.

File: blackjax/inference/integrators.py
Comment (rlouf) [reply blackjax/inference/integrators.py]: Pushed.

File: blackjax/inference/proposals.py
Comment (rlouf) [reply blackjax/inference/proposals.py]: I pushed the modification, but leaving the discussion open in case you want to add something.

File: blackjax/inference/integrators.py
Comment (rlouf) [reply blackjax/inference/integrators.py]: Pushed.

File: blackjax/inference/integrators.py
Comment (rlouf) [reply blackjax/inference/integrators.py]: Pushed.

File: blackjax/inference/metrics.py
Comment (rlouf) [reply blackjax/inference/metrics.py]: Pushed.

File: blackjax/inference/proposals.py
Comment (rlouf) [reply blackjax/inference/proposals.py:65]: :+1: to keep it, :-1: to get rid of it.

Comment (fehiepsi) [reply blackjax/inference/proposals.py:65]: It depends on your usage case. I was worried that if step_size=1e-6 and trajectory length is 1 (which is popular for tricky posterior I believe) then this will hold 1e6 samples in memory.

File: blackjax/inference/integrators.py
Comment (fehiepsi) [reply blackjax/inference/integrators.py:59]: I think so too.

File: blackjax/inference/proposals.py
Comment (rlouf) [reply blackjax/inference/proposals.py:65]: I never thought that someone could integrate the trajectory 1e6 times in 1 HMC step! If that's common we indeed better make sure the memory doesn't explode doing so. I'll implement the change, we can always revert the commit if someone knows for sure that XLA makes some wicked optimization when the states are later discarded.

**Comment by rlouf** [comment]
I finished implementing the proposed changes. @fehiepsi suggested a performance micro-optimization that we should keep in mind and experiment with later.

@twiecki @neerajprad @fehiepsi Unless you have other comments I will merge this PR tomorrow pm (Central European Time) and rebase the NUTS branch on this one.

File: blackjax/inference/proposals.py
Comment (fehiepsi) [reply blackjax/inference/proposals.py]: Haha, I believe no one wants to do so, but adaptive scheme might be not smart (if accept probs in initial steps are small, step size will be dropped pretty fast). ;)

File: blackjax/hmc.py
Comment (neerajprad) [reply blackjax/hmc.py:87]: Sounds good, I was only asking since I saw window adaptation being referred here but wasn't sure about the plan. It will be good to provide some adaptation scheme for sure. If that requires any changes to the interface, we can discuss that later.

File: blackjax/inference/base.py
Comment (neerajprad) [reply blackjax/inference/base.py:179]: I see, that makes sense. If a proposal is rejected, we have both the earlier state and the proposed state to debug why it was rejected.

**Review by neerajprad — APPROVED** [review]

File: blackjax/inference/proposals.py
Comment (rlouf) [reply blackjax/inference/proposals.py]: True for jitting, but then we don't have to return the trajectory since we encapsulate the information in `ProposalInfo`. We will add adaptive schemes, so I think it's better to drop it for now :)

**Comment by rlouf** [comment]
Thanks everyone for reviewing the code! Merging :champagne:

**Pull request merged**
